src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs

```diff
@@ -24,37 +20,67 @@ private DriverStateTable(ImmutableDictionary<object, IStateTable> tables)
             _tables = tables;
         }
 
-        internal NodeStateTable<T> GetStateTable<T>(IIncrementalGeneratorNode<T> input) => _tables.ContainsKey(input) ? (NodeStateTable<T>)_tables[input] : NodeStateTable<T>.Empty;
-
-        internal DriverStateTable SetStateTable<T>(IIncrementalGeneratorNode<T> input, NodeStateTable<T> table) => new DriverStateTable(_tables.SetItem(input, table));
+        public NodeStateTable<T> GetStateTableOrEmpty<T>(object input)
+        {
+            if (_tables.TryGetValue(input, out var result))
+            {
+                return (NodeStateTable<T>)result;
+            }
+            return NodeStateTable<T>.Empty;
+        }
 
         public sealed class Builder
         {
             private readonly ImmutableDictionary<object, IStateTable>.Builder _tableBuilder = ImmutableDictionary.CreateBuilder<object, IStateTable>();
-
+            private readonly ImmutableArray<ISyntaxInputNode> _syntaxInputNodes;
             private readonly DriverStateTable _previousTable;
-
             private readonly CancellationToken _cancellationToken;
 
-            public Builder(DriverStateTable previousTable, CancellationToken cancellationToken = default)
-            {
-                _previousTable = previousTable;
-                _cancellationToken = cancellationToken;
-            }
+            internal GeneratorDriverState DriverState { get; }
 
-            public void SetTable<T>(IIncrementalGeneratorNode<T> source, NodeStateTable<T> table)
-            {
-                _tableBuilder[source] = table;
-            }
+            public Compilation Compilation { get; }
 
-            public void AddInput<T>(InputNode<T> source, T value)
+            public Builder(Compilation compilation, GeneratorDriverState driverState, ImmutableArray<ISyntaxInputNode> syntaxInputNodes, CancellationToken cancellationToken = default)
             {
-                _tableBuilder[source] = source.CreateInputTable(_previousTable.GetStateTable(source), value);
+                Compilation = compilation;
+                DriverState = driverState;
+                _previousTable = driverState.StateTable;
+                _syntaxInputNodes = syntaxInputNodes;
+                _cancellationToken = cancellationToken;
             }
 
-            public void AddInput<T>(InputNode<T> source, IEnumerable<T> value)
+            public IStateTable GetSyntaxInputTable(ISyntaxInputNode syntaxTransform)
             {
-                _tableBuilder[source] = source.CreateInputTable(_previousTable.GetStateTable(source), value);
+                // when we don't have a value for this node, we update all the syntax inputs at once
+                if (!_tableBuilder.ContainsKey(syntaxTransform))
+                {
+                    // get a builder for each input node
+                    var builders = ArrayBuilder<ISyntaxInputBuilder>.GetInstance(_syntaxInputNodes.Length);
+                    foreach (var node in _syntaxInputNodes)
```

Comment:
>_syntaxInputNodes [](http://example.com/codeflow?start=41&length=17)

Is the requirement that `syntaxTransform` is one of the nodes in `_syntaxInputNodes`? If so, perhaps add an assert at the start of the method.